### DISCUSSION TITLE
Is there a way to intercept HTTP requests made by Craft and modify the headers?

### DISCUSSION THREAD

**Discussion by jamesmacwhite** [craftcms/cms#16126] [Q&A] [answered]
I have a Feed Me feed which provides image URLs on a CDN path, the issue is the CDN is using Cloudflare and our server environment which happens to be Craft Cloud is being caught Cloudflare's bot/fingerprinting detection and the image URLs when being attempted to be downloaded are getting a 403.

I've tested locally and found by specifying a User-Agent the request is successful. Unfortunately, for asset downloading in Feed Me there is no way to add/modify HTTP headers with Curl or fopen. I raised an issue on the Feed Me GitHub: https://github.com/craftcms/feed-me/issues/1538

I was however wondering if there any way to use an event/app.web.php to intercept requests and add/modify headers depending on a condition i.e. hostname as a workaround. Feed Me I don't think provides any events specifically for asset fetching, so it's not really able to manipulate it, so I was wondering if I could do it somehow at the higher level.

**Comment by brandonkelly** [comment, marked as answer]
You can [modify the config](https://craftcms.com/docs/5.x/configure.html#guzzle) that Craft uses to create Guzzle (HTTP) clients via `config/guzzle.php`:

```php
<?php

return [
    'headers' => [
        'User-Agent' => '...',
    ],
];
```

Looks like it’s not documented (on it), but Feed Me [supports](https://github.com/craftcms/feed-me/blob/153e1eeb66981a6c695dce0f021cb4ff771692a7/src/services/Service.php#L49) setting a `clientOptions` array on your feed configs from `config/feed-me.php`, which will get merged into the main Guzzle config:

```php
<?php

return [
    '*' => [
        'clientOptions' => [
            'headers' => [
                'User-Agent' => '...',
            ],
        ],
    ],
];
```

**Reply by jamesmacwhite** [reply]
Hi Brandon. I'm not sure these settings will apply in this case, because the AssetHelper class uses either fopen or curl, not Guzzle and wouldn't apply these settings. For the main Feed model with the Feed URL it does apply, but the asset fetching calls a different client I believe.
